lib/dialects/mysql/query.js

```diff
@@ -33,32 +33,25 @@ class Query extends AbstractQuery {
     //do we need benchmark for this query execution
     const showWarnings = this.sequelize.options.showWarnings || options.showWarnings;
 
-    const complete = this._logQuery(sql, debug, parameters);
-
-    const results = await new Promise((resolve, reject) => {
-      const handler = (err, results) => {
-        complete();
+    const query = parameters && parameters.length
+      ? new Promise((resolve, reject) => connection.execute(sql, parameters, (error, result) => error ? reject(error) : resolve(result)).setMaxListeners(100))
+      : new Promise((resolve, reject) => connection.query({ sql }, (error, result) => error ? reject(error) : resolve(result)).setMaxListeners(100));
 
-        if (err) {
-          // MySQL automatically rolls-back transactions in the event of a deadlock
-          if (options.transaction && err.errno === 1213) {
-            options.transaction.finished = 'rollback';
-          }
-          err.sql = sql;
-          err.parameters = parameters;
+    const complete = this._logQuery(sql, debug, parameters);
 
-          reject(this.formatError(err));
-        } else {
-          resolve(results);
-        }
-      };
-      if (parameters) {
-        debug('parameters(%j)', parameters);
-        connection.execute(sql, parameters, handler).setMaxListeners(100);
-      } else {
-        connection.query({ sql }, handler).setMaxListeners(100);
+    const results = await query.catch(err => {
+      // MySQL automatically rolls-back transactions in the event of a deadlock
+      if (options.transaction && err.errno === 1213) {
+        options.transaction.finished = 'rollback';
       }
+      err.sql = sql;
+      err.parameters = parameters;
+
+      throw this.formatError(err);
     });
+
+    complete();
```

Comment:
Nice, I think it makes sense for `complete()` to be called only if no error happens. Good change :+1: